devtools/etdump/buffer_data_sink.cpp

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <executorch/devtools/etdump/buffer_data_sink.h>
+#include <executorch/devtools/etdump/utils.h>
+
+using ::executorch::runtime::Error;
+using ::executorch::runtime::Result;
+
+namespace executorch {
+namespace etdump {
+
+Result<size_t> BufferDataSink::write(const void* ptr, size_t length) {
+  if (length == 0) {
+    return static_cast<size_t>(-1);
+  }
+
+  uint8_t* last_data_end = debug_buffer_.data() + offset_;
+
+  // The beginning of the next data blob must be aligned to the alignment
+  uint8_t* cur_data_begin = internal::align_pointer(last_data_end, alighment_);
+  uint8_t* cur_data_end = cur_data_begin + length;
+
+  // Raise OutOfResources error if the end of current data blob is out of range.
+  if (cur_data_end > debug_buffer_.data() + debug_buffer_.size()) {
+    return Error::OutOfResources;
+  }
+
```

Comment:
Please bring back the memset to zero out the padding, and add a test that checks for zero-value pad bytes